arbitrum-client/src/client/config.rs

```diff
@@ -0,0 +1,87 @@
+//! The configuration options for the Arbitrum client, along with
+//! convenience methods for converting them into richer types that the
+//! client works with.
+
+use std::{error::Error, fmt::Display, str::FromStr, sync::Arc};
+
+use ethers::{
+    core::k256::ecdsa::SigningKey,
+    middleware::SignerMiddleware,
+    providers::{Http, Middleware, Provider},
+    signers::{LocalWallet, Signer, Wallet},
+    types::Address,
+};
+
+use crate::abi::DarkpoolContract;
+
+pub struct ArbitrumClientConfig {
+    /// The address of the darkpool proxy contract.
+    ///
+    /// This is the main entrypoint to interaction with the darkpool.
+    pub proxy_address: String,
+    /// The address of the darkpool implementation contract.
+    ///
+    /// This is used to filter for events emitted by the darkpool.
+    pub implementation_address: String,
+    /// The HTTP-addressable URL of the Arbitrum RPC endpoint
+    pub rpc_url: String,
+    /// The private key of the account to use for signing transactions
+    pub arb_priv_key: String,
+}
+
+#[derive(Clone, Debug)]
+pub enum ArbitrumClientConfigError {
+    RpcClientInitialization(String),
+    AddressParsing(String),
+}
+
+impl Display for ArbitrumClientConfigError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+impl Error for ArbitrumClientConfigError {}
+
+pub type SignerHttpProvider = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
+
+impl ArbitrumClientConfig {
+    async fn get_rpc_client(&self) -> Result<Arc<SignerHttpProvider>, ArbitrumClientConfigError> {
+        let provider = Provider::<Http>::try_from(&self.rpc_url)
+            .map_err(|e| ArbitrumClientConfigError::RpcClientInitialization(e.to_string()))?;
+
+        let wallet = LocalWallet::from_str(&self.arb_priv_key)
+            .map_err(|e| ArbitrumClientConfigError::RpcClientInitialization(e.to_string()))?;
+
+        let chain_id = provider
+            .get_chainid()
+            .await
+            .map_err(|e| ArbitrumClientConfigError::RpcClientInitialization(e.to_string()))?
+            .as_u64();
+
+        let rpc_client = Arc::new(SignerMiddleware::new(
+            provider,
+            wallet.clone().with_chain_id(chain_id),
+        ));
+
+        Ok(rpc_client)
+    }
+
+    fn get_proxy_address(&self) -> Result<Address, ArbitrumClientConfigError> {
+        Address::from_str(&self.proxy_address)
+            .map_err(|e| ArbitrumClientConfigError::AddressParsing(e.to_string()))
+    }
+
+    pub fn get_implementation_address(&self) -> Result<Address, ArbitrumClientConfigError> {
+        Address::from_str(&self.implementation_address)
+            .map_err(|e| ArbitrumClientConfigError::AddressParsing(e.to_string()))
+    }
```

Comment:
Do we ever need these fields as strings? Probably best to store them as `Address` to avoid expensive string parsing when needed

Reply:
I'm just thinking ahead to how this config object will be constructed - at the end of the day, it will come from CLI args which come in as strings. So at some point we will need to do string parsing, and it makes sense to me to keep this config struct lightweight to instantiate, and derive rich types for the full client struct.

Fwiw this is also the precedent set in the `StarknetClientConfig`

Reply:
Oh gotcha, if it's only parsed once then it doesn't matter